### ISSUE TITLE
StackingEstimator.transform can include NaNs in the transformed X and cause TPOTBase._summary_of_best_pipeline to fail

### ISSUE THREAD

**Issue by JoshHeitzman** [EpistasisLab/tpot#893]
With version 0.9.6 I observed an exception about NaN/Infinite values during the execution of TPOTBase._summary_of_best_pipeline as it was iterating to fill in self.pareto_front_fitted_pipelines_, but after triple checking my input features and labels I couldn't locate any nan or inf values.  But I did find that return value of this statement in Pipeline._fit:

            Xt, fitted_transformer = fit_transform_one_cached(
                cloned_transformer, Xt, y, None,
                message_clsname='Pipeline',
                message=self._log_message(step_idx),
                **fit_params_steps[name])

had NaNs in it although the input value did not.  From there I traced it to the StackingEstimator.transform function and added a check to its input and return value to verify that it is the source of the NaN values inspite of not being passed any NaN values.

Unrelated to this I had modified gp_deap._wrapped_cross_val_score to print out the the exception before return -float('inf') but I wasn't seeing this get hit for this particular pipeline so it seems to have made it through the initial fit without errors only to raise an error in the fit in TPOTBase._summary_of_best_pipeline.

I've worked around this by simply disabling the second fitting of models to fill in the self.pareto_front_fitted_pipelines_ since I'm not using that and it doesn't seem to be required for TPOT to operate.

**Comment by weixuanfu** [comment]
Thanks for tracking this errors. 

I think it could be a bug in `StackingEstimator` for stacking NAN/infinite prediction/predication probabilities (maybe related to [this issue](https://stackoverflow.com/questions/49525618/sklearns-predict-proba-returns-infinite-probabilties)). But it is strange that the pipeline passed during CV but did not work in `fit` step in `summary_of_best_pipeline`. Could you please let us know all the steps in this pipeline for reproducing this error?

**Comment by JoshHeitzman** [comment]
When the exception is raised, up the stack in _summary_of_best_pipeline I can get the following print of the steps for the pipeline being processed:

for i, step in enumerate(self._toolbox.compile(expr=pipeline).steps):
    print(i, step)

0 ('stackingestimator', StackingEstimator(estimator=LogisticRegression(C=0.01, class_weight=None,
                                               dual=True, fit_intercept=True,
                                               intercept_scaling=1,
                                               l1_ratio=None, max_iter=100,
                                               multi_class='warn', n_jobs=None,
                                               penalty='l2', random_state=None,
                                               solver='warn', tol=0.0001,
                                               verbose=0, warm_start=False)))
1 ('bernoullinb', BernoulliNB(alpha=0.001, binarize=0.0, class_prior=None, fit_prior=False))

**Comment by JoshHeitzman** [comment]
@weixuanfu I believe I've figured out what is allowing for exceptions to be thrown in _summary_of_best_pipeline when the pipeline didn't fail during the search.  In _summary_of_best_pipeline the complete set of data is used to fit the pipeline, where as during search its only a subset(s) of the data as some data is reserved for validation.  I ran into another case were it was the fit of the best pipeline that raised an exception in _summary_of_best_pipeline.  In this case the StackingEstimator wasn't involved.  Instead sklearn.feature_selection.SelectFwe was being used and it was doing a transform that resulted in theere being zero feature columns, butut this only happened when it was fit to the entire data set and not to the subset used for training.

My current workaround it is to modify gp_deap._wrapped_cross_val_score to do:
clone(sklearn_pipeline).fit(features, target)

just before the loop on cv_iter, so that the problem is detected at a point where -inf can be returned to cause the pipeline to be ignored in favor other pipelines that won't fail on the final fit to the full data set.

**Comment by weixuanfu** [comment]
@JoshHeitzman Thank for checking this issue again. Based on your findings, I think there are two problems here:

1. `StackingEstimator` stacked NaN/infinite values so that the best pipeline failed when fitting it on entire dataset. This issue can be solved by adding NaN/infinte checking/imputation into `StackingEstimator`

2. `SelectFwe` returned a empty X when fitting it on entire dataset, which I think it should not happen very frequently . For fixing this issue, maybe we should change `_update_top_pipeline` function to checking if the pipeline can be fitted on entire dataset instead of checking it during search.

I am wondering what is the `cv` setting in your case. It seems that the training subsets in CV was very different with entire dataset.

**Comment by JoshHeitzman** [comment]
@weixuanfu In this case the cv has been provided as class instance whose split method yields one pairs of training and test indexes.  The indexes are the first (in time) 80% of the data and test set is the last 20% of the data.

It seems that _update_top_pipeline is only called immediately before _summary_of_best_pipeline except when checkpoints are being written out, so it seems like it would be possible that a check in _update_top_pipeline only after the search could find that there is no valid pipeline to replace the top one with if it should be found invalid there.

In my overnight run I encountered another failure fitting the top pipeline in _summary_of_best_pipeline even with my workaround in place on somewhat different data then that workaround initially allowed running through cleanly.  Investigating.

**Comment by JoshHeitzman** [comment]
@weixuanfu I've also opened https://github.com/EpistasisLab/tpot/issues/894 as that appears to be an independent issue, although it is generally contributing to allowing an exception to be raised during the final fit of the top model.  In the third comment on this thread, that random_state is also being provided a value of None although the classifier was created with a random_state of 1 in that case as well.

**Comment by weixuanfu** [comment]
The issue has been fixed in new version of TPOT (v0.11.0). Please feel free to reopen it if you have any questions or suggestions.
